### ISSUE TITLE
Support for `std.StringArrayHashMap` in the type to `std.json.parse` (i.e. runtime-known keys in JSON data but typed)

### ISSUE THREAD

**Issue by wooster0** [ziglang/zig#11712]
I propose adding support for hash maps/objects inside the type given to `std.json.parse`'s `comptime T: type`, even if it's only string hash maps (`std.StringArrayHashMap`/`std.StringHashMap`), for runtime-known (string) keys in an object. I think this is one rather important feature that the inbuilt `std.json` should definitely support.

The point of this proposal is that **we have to use `std.json.Parser` and can't use `std.json.parse`** if our **JSON data contains a single runtime-known key.**

I have collected some preliminary feedback in [this](https://github.com/ziglang/zig/issues/8660#issuecomment-1136496886) comment and I was suggested to use `std.json.Parser` which, yes, can be used as a workaround (and I'm using it as a workaround currently) because it allows you to parse *anything* without specifying any types. However, I have the types. I know what types my data will be and I want it to be typed. Having to suddenly change your whole code to use `std.json.Parser` instead of `std.json.parse` just because you have a single string field anywhere in your JSON object that is runtime-known is frustrating. You lose all type information, all guarantees about your data, and have to check manually at runtime as you access the object whether or not the data exists, as opposed to doing this validation at parse time. I want my data to be typed.

Here's a comparison, parsing the same data in two ways; in the latter we will pretend that `json_string` contains a key we don't know the name of:
```zig
const std = @import("std");

// we will parse this twice in two different ways below
const json_string =
    \\{
    \\    "hello": {
    \\        "world": "hello"
    \\    }
    \\}
;

pub fn main() !void {
    var arena_child = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    var arena = std.heap.loggingAllocator(arena_child.allocator());
    const arena_instance = arena.allocator();

    //
    // the following we can only do if we know the exact name of the key at compile time!
    //
    const json_type = struct { hello: struct { world: []const u8 } };
    std.log.debug("std.json.parse: efficient and typed", .{});
    var tokens = std.json.TokenStream.init(json_string);
    // All data is validated here and we can handle any error with a single `try`.
    const result = try std.json.parse(json_type, &tokens, .{ .allocator = arena_instance });
    // Now you've got all benefits types have to offer
    std.log.debug("{s}\n", .{result.hello.world});

    // --------------------------------------

    //
    // for the following let's pretend we don't know of either the "hello" or "world" keys
    // at compile time. the following is what we have to do if we have just a *single* runtime-known key!
    //
    std.log.debug("std.json.Parser: none of the benefits and all of the problems", .{});
    var parser = std.json.Parser.init(arena_instance, false);
    const value_tree = try parser.parse(json_string);
    // Validation has to be done post-parsing
    if (value_tree.root == .Object) {
        if (value_tree.root.Object.get("hello")) |something| {
            if (something == .Object) {
                if (something.Object.get("world")) |something_else| {
                    if (something_else == .String) {
                        std.log.debug("{s}", .{something_else.String});
                    } else {
                        // our data is wrong
                    }
                } else {
                    // our data is wrong
                }
            } else {
                // our data is wrong
            }
        } else {
            // our data is wrong
        }
    } else {
        // our data is wrong
    }
}
```
```
$ zig run x.zig
debug: std.json.parse: efficient and typed
debug: alloc - success - len: 5, ptr_align: 0
debug: hello

debug: std.json.Parser: none of the benefits and all of the problems
debug: alloc - success - len: 448, ptr_align: 3
debug: alloc - success - len: 984, ptr_align: 3
error: expand - failure - 984 to 1240, buf_align: 0
debug: alloc - success - len: 2424, ptr_align: 3
debug: hello
```
Clearly `std.json.parse` is superior because it's faster and uses less memory.
This is another reason `std.json.Parser` is a very unsatisfactory permanent solution.

Can we please also add support for at least string hash maps in the type to `std.json.parse`? At least one of `std.StringArrayHashMap` or `std.StringHashMap` should be supported.

## Prior art

https://github.com/ziglibs/tres supports this. But you can easily come across this kind of JSON in the wild so `std.json` should definitely have it built-in. tres can be used as a reference.

---

Related: #8660

**Comment by courtier** [comment]
This would be great to have indeed.

**Comment by r00ster91** [comment]
Yeah, I'm currently dealing with JSON where a few keys are runtime-known and just because I have a single runtime-known key I have to deal with `std.json.ValueTree` and write a jungle of `if`s and `for`s and my currently deepest indentation there is 48 spaces. The code is slow to read, cumbersome to write, and it's very easy to make mistakes. All this would be avoided if the type for `std.json.parse` allowed runtime-known keys to be specified.

**Comment by thejoshwolfe** [comment]
`std.json.ArrayHashMap` added in https://github.com/ziglang/zig/pull/16366 . Available in version Zig 0.11.0.
